tbot_contrib/locking.py

```diff
@@ -0,0 +1,299 @@
+        shell_pid = self.env("$") if self.lock_checkpid else "0"
+        tempfile.write_text(f"{time_str} {shell_pid}\n")
+        return tempfile
+
+    def _lock_handle_expiry(self, lockfile: linux.Path) -> bool:
+        locklock = self.lock_dir / "lock-lock"
+        # open locklock as file descriptor
+        self.exec0("exec", linux.Raw(f"{self.lock_fd}>"), locklock)
+        # flock on the file descriptor
+        retval, result = self.exec("flock", str(self.lock_fd))
+        print(retval, result)
+        # Read the lock, perhaps it is stale...
+        retval, result = self.exec("cat", lockfile)
+        if retval != 0:
+            return False
+        repeat = False
+        results = result.split()
+        # Check if the lock can expire, and whether it has expired.
+        if (int(results[0]) != 0) and (int(results[0]) < int(time.time())):
+            # Check whether the lock PID is still in use by a bash instance
+            retval, result = self.exec("ps", "-ocomm=", results[1])
+            if (retval == 0) and (result.strip() == "bash"):
+                print(f"Expired lock {lockfile} PID still appears valid!")
+            else:
+                self.exec("rm", lockfile)  # Delete stale Lock
+                self._active_locks.discard(lockfile)
+                print(f"Lock {lockfile} expired, deleted.")
+                repeat = True  # The stale lock may be available next time round.
+        # unflock the file descriptor
+        self.exec0("flock", "-u", str(self.lock_fd))
+        # close the file descriptor
+        self.exec0("exec", linux.Raw(f"{self.lock_fd}>&-"))
+        return repeat
+
+    def request_machine_lock(
+        self, name: str, *, expiry: typing.Optional[int] = None
+    ) -> bool:
+        if not self.lock_dir.is_dir():
+            self.exec0("mkdir", self.lock_dir)
+            self.exec0("chmod", "0777", self.lock_dir)
+        lockfile = self.lock_dir / name
+        tempfile = self._lock_make_temp(expiry)
+        success = False
+        repeat = True
+        while repeat:  # do-while
+            repeat = False
+            print(f"Try locking on {lockfile}...")
+            if self.test("ln", tempfile, lockfile):  # hard-linking is atomic
+                success = True
+                repeat = False  # We got a lock, we're done
+            else:  # No lock this time, handle expiry. This may race!
+                repeat = self._lock_handle_expiry(lockfile)
+        self.exec("rm", tempfile)  # Delete the temp file
+        if success:  # insert lock into set
+            self._active_locks.add(lockfile)
+        return success
+
+    # The problem here is, if someone else reclaimed and reused a lock we held,
+    # we may now remove it here...
+    def release_machine_lock(self, name: str) -> None:
+        lockfile = self.lock_dir / name
+        if lockfile in self._active_locks:
+            self._active_locks.remove(lockfile)
+            self.exec("rm", "-f", lockfile)
```

Comment:
Looking at this more, I think this line should also be dropped:

```python
if lockfile in self._active_locks:
```

If someone calls `release_machine_lock()` for a lock that is not held, this indicates bad downstream code and thus we should fail with noise instead of silently ignoring it.